### ISSUE TITLE
Using authenticated routes with mounted engines causes a redirect loop

### ISSUE THREAD

**Issue by ajhodges** [plataformatec/devise#3613]
Using Rails 4.2.1, Devise 3.5.1, Warden 1.2.3

This is a regression in 3.5.x. I have `rails_admin` and `sidekiq` mounted under an authenticated admin route in my app like so:

    devise_for :admin

    authenticate :admin do
      scope '/admin' do
        #reqsue admin panel
        mount Sidekiq::Web => "/sidekiq"

        #rails_admin admin panel
        mount RailsAdmin::Engine => '/', as: 'rails_admin'
      end
    end

The behavior with devise 3.4.1 is that if I navigate to `/admin` or `/admin/sidekiq` and I am not authenticated as an `Admin`, I get redirected to `/admin/sign_in`. After upgrading to 3.5.1, I am redirected to `/admin/admin/sign_in` and `/admin/sidekiq/admin/sign_in` respectfully. This causes a redirect loop.

**Comment by benjreinhart** [comment]
I have a very similar issue, also with Rails 4.2.1, Devise 3.5.1 and Warden 1.2.3.

I have the following setup:

```ruby
devise_for :users, skip: [:sessions, :registrations]

as :user do
  get 'login' => 'devise/sessions#new', :as => :new_user_session
end

namespace :admin do
  authenticate :user, -> (u) { u.is_admin? } do
    mount Sidekiq::Web => '/sidekiq'
  end
end
```

The desired functionality is that when a user is not logged in, they should be redirected to `/login`. In Devise 3.4.1 this is the case. However, in 3.5.1 they are redirected to `/admin/sidekiq/login`.

If I move the `authenticate` call out of the `namespace :admin` block but mount `Sidekiq` to `/admin/sidekiq`, I also get a redirect to `/admin/sidekiq/login` when a user is not logged in:

```ruby
authenticate :user, -> (u) { u.is_admin? } do
  mount Sidekiq::Web => '/admin/sidekiq'
end
```

This redirects to `/admin/sidekiq/login`.

Lastly, if I change the above to be mounted at just `/sidekiq`, I get a redirect to `/sidekiq/login` when user is not logged in:

```ruby
authenticate :user, -> (u) { u.is_admin? } do
  mount Sidekiq::Web => '/sidekiq'
end
```

This redirects to `/sidekiq/login`.

**Comment by siruguri** [comment]
I'm wondering if this is a feature, not a bug. The CHANGELOG describes the change as 'Use router_name from scope if one is available to support isolated engines' in the bugfix section.

I don't know the insides of Devise enough to say, but I'm guessing there's a method somewhere, or an option to `authenticate`, that you can use to tell the router that your mounted engines want to authenticate using scopes outside the engine. Or you could insert a route higher up in the routes file that catches the mounted engine's redirect and sends it to your "top" app. I did this for example:

      get '/rails_admin/users/sign_in' => redirect('/users/sign_in')


I have never set `relative_url_root` but I'd guess all relative roots would then move to the top level root, which isn't what you'd want inside your mounted engines, yes?

**Comment by stevenharman** [comment]
This was fixed with aa675f7f6611de5d40114aa98153f3f0c0b3e8d5, as mentioned in #3643.

**Comment by aYaremchuk** [comment]
rails (5.0.0.1),  devise (4.0.3), warden (1.2.6)

routes.rb:
`devise_for :users

 authenticate :user do
    mount Sidekiq::Web => '/sidekiq'
  end
`

When user is not authorized I get redirect loop error.
I'm trying visit `http://localhost:3000/sidekiq` but I get redirect to `http://localhost:3000/sidekiq/users/sign_in` instead of `http://localhost:3000/users/sign_in`.
